File: web/src/lib/linkUrl.ts
/**
 * A manually-typed or AI-extracted link often has no "https://" — this
 * adds one so it's actually clickable (browsers resolve a bare
 * "instagram.com/..." href as a relative path, not the address it looks
 * like) and so hostname parsing below actually works.
 */
export function normalizeLinkHref(url: string): string {
  return /^[a-z][a-z0-9+.-]*:\/\//i.test(url) ? url : `https://${url}`;
}

/** Whether a place's manually-entered/AI-extracted link points at
 *  Instagram — used to label it "Instagram" instead of the generic
 *  "Website" wherever it's shown. */
export function isInstagramLink(url: string): boolean {
  try {
    return new URL(normalizeLinkHref(url)).hostname.replace(/^www\./, "") === "instagram.com";
  } catch {
    return false;
  }
}

File: Peragra/Services/LinkURL.swift
import Foundation

/// Shared normalization for a manually-entered or AI-extracted
/// Instagram/website link — a person (or a model reading a caption) often
/// gives one without "https://" (e.g. "instagram.com/x"), which
/// URL(string:) would otherwise treat as a relative path rather than the
/// address it looks like.
enum LinkURL {
    static func normalize(_ raw: String) -> URL? {
        guard !raw.isEmpty else { return nil }
        if raw.range(of: "^[a-zA-Z][a-zA-Z0-9+.-]*://", options: .regularExpression) != nil {
            return URL(string: raw)
        }
        return URL(string: "https://\(raw)")
    }

    /// Whether a link points at Instagram — used to label it "Instagram"
    /// instead of the generic "Website" wherever it's shown.
    static func isInstagram(_ raw: String) -> Bool {
        normalize(raw)?.host?.replacingOccurrences(of: "^www\\.", with: "", options: .regularExpression) == "instagram.com"
    }
}
